Datetime command

tiomny opened the pull request:
Introduced new "Datetime" command.
Usage:
`Datetime[,YYYY-MM-DD[,hh:mm:ss]]`
Examples:
`Datetime,2020-02-29,00:23:00`
`Datetime,2020-02-29` - set node time "2020-02-29 00:00:00"
`Datetime` - get current node time

Important note: "Use NTP" option should be set.
Resolved #2677

tiomny committed:
SetDatetime command

giig1967g commented:
Hi,
Please wait until #2778 is merged as the ExecuteCommand syntax will change.

TD-er commented on src/src/Commands/Time.cpp:
```diff
@@ -54,3 +55,36 @@ String Command_DST(struct EventStruct *event, const char* Line)
+		sysTime = makeTime(tm);
+		
+	}else  {
+		serialPrintln();
```
Why do you write to the Serial port here?

tiomny replied:
> 
> 
> Why do you write to the Serial port here?

I copied code from a similar function in Time.cpp - perhaps, that's not needed, as well in other functions in this module.

TD-er replied:
There is only a limited number of commands (for debug purposes mainly) that output to the serial port.
For example meminfo and meminfodetail.
And I guess even those are no longer needed to output to the serial port.

TD-er commented on src/src/Commands/Time.h:
```diff
@@ -3,9 +3,12 @@
 
 class String;
 
+#include <time.h>
```
Why is this include needed here? 
Only include in .h files if you need to have some struct as parameter or return value (or member of a class)
Sometimes you can even get away with not including in a .h file by forward declaration of the class you need.
But here you don't need any of those in the .h file, so include it in the .cpp where you actually need it.

tiomny replied:
Removed.

tiomny replied:
Removed.

TD-er commented on src/TimeESPeasy.ino:
```diff
@@ -3,19 +3,12 @@
 struct tm tm;
 uint32_t  syncInterval = 3600;       // time sync will be attempted after this many seconds
-double    sysTime = 0.0;             // Use high resolution double to get better sync between nodes when using NTP
+// double    sysTime = 0.0;             // Use high resolution double to get better sync between nodes when using NTP
```
Why do you remove `sysTime`?

tiomny replied:
> 
> 
> Why do you remove `sysTime`?

I moved it to TimeESPeasy.cpp to make it available from Time.cpp.

tiomny replied:
Reverted back.

TD-er replied:
OK, maybe we should then get rid of TimeESPeasy.ino and move the global defined variables to a separate .h and .cpp file in the Globals directory.

Then global defined variables must in the .h file have the `extend` keyword and only declared once in their .cpp file.
See other .h/.cpp examples in the Globals directory.

The tm struct should then be defined in DataStructs directory and its global variable in the Globals, as it is a global defined variable to keep track of the current time.

The reason I'm transitioning .ino files into their .h/,cpp is mainly because of build issues we're seeing with .ino files.
Also it does make it clear where things are defined, when they are included, etc. So then we can use the tools like cpp-check and the compiler to check our code, which is much harder on .ino files.

tiomny replied:
That sounds reasonable, but isn't that better to make refactoring separately from other stuff?

TD-er replied:
Well you are moving things around... So we're already mixing it.
Sometimes adding features and refactoring are not that black-and-white (well it is more rare it is clearly a different process)

tiomny committed:
Rolled back extra changes

tiomny committed:
Removed extra includes

TD-er commented:
I noticed you do merge your own PR branch into your "mega" branch.
That's a strange move as you normally would make a branch for a new feature, then push commits to it and offer that branch as a pull request.
If you then would like to work again on another branch, your exiting pull request will not be polluted with these new commits.
On the other hand if you later want to add something to the pull request, you checkout that branch and push commits to it. as soon as you push your commits to your own origin/<branch_name> it will be added to this PR.

So keep that in mind if you plan to push commits made on your mega branch.

tiomny commented:
> 
> 
> I noticed you do merge your own PR branch into your "mega" branch.
> That's a strange move as you normally would make a branch for a new feature, then push commits to it and offer that branch as a pull request.
> If you then would like to work again on another branch, your exiting pull request will not be polluted with these new commits.
> On the other hand if you later want to add something to the pull request, you checkout that branch and push commits to it. as soon as you push your commits to your own origin/<branch_name> it will be added to this PR.
> 
> So keep that in mind if you plan to push commits made on your mega branch.

Thanks, I will keep that in mind. I'm more TFS person.

tiomny committed:
Removed serialPrintln

TD-er commented:
> I'm more TFS person.

Well I guess Microsoft tools also would like to provide some structure in development, when working on it with > 1 person?
I never used it, so can't say for sure.

If you jump back and forth between features or branch states, you will make it much much harder for yourself if you want to delay merging a feature or roll back of a feature.
Been there, done that, learned the hard way.

tiomny commented:
> 
> 
> > I'm more TFS person.
> 
> Well I guess Microsoft tools also would like to provide some structure in development, when working on it with > 1 person?
> I never used it, so can't say for sure.
> 
> If you jump back and forth between features or branch states, you will make it much much harder for yourself if you want to delay merging a feature or roll back of a feature.
> Been there, done that, learned the hard way.

TFS provides much more convenient way of merging branches. Usually we merge only particular changesets.

uzi18 commented:
One fact is always true - small change is easier to analyse.

TD-er commented:
Just a notice, as I will move the time related functions into a .h/.cpp file.

TD-er commented:
Can you update (or maybe redo) this PR, based on the current state of the mega branch?
The time related code has moved a bit.

tiomny commented:
I will.

tiomny committed:
Remade using current code base

tiomny committed:
Removed extra linefeed

tiomny committed:
Added missed linefeeds

tiomny commented:
I made a rebase and changed code in PR accordingly.

TD-er commented:
Looks clean.
Will test it.

uzi18 commented:
If we provide SetDateTime, so maybe we can try also Get it ;)

TD-er commented:
> If we provide SetDateTime, so maybe we can try also Get it ;)

Pfft give people a finger and they take the full hand ;)
There is only one thing I can say about it...: "Why didn't I think of that?" :)

tiomny commented:
> If we provide SetDateTime, so maybe we can try also Get it ;)

SetDatetime without parameters returns current node time. Besides, current time could be taken from <ip>/json.

TD-er commented:
Ah that's true indeed.
Forgot that I had made it into a get/set function.
But that also implies the name of the command should not include "get" or "set", as is with the other commands.

And I was thinking, maybe the functionality to set it should be in the node_time object.
Now you're doing a lot outside that class and access a member which should not be public anyway (but that would involve too much changes for now) to set the time.
But we'll leave it for now.

Could you just rename the command to `DateTime`? And add some documentation to it as well.

tiomny commented:
> 
> Could you just rename the command to `DateTime`? And add some documentation to it as well.

Sure.

tiomny committed:
Change SetDatetime to Datetime command

TD-er committed:
[DateTime] Documented command + few fixes

- Argument number was incorrect
- did not set time source.  (so you had to set NTP which defeats the purpose of this command
- Added documentation
- Moved the command call to the correct first letter of the alphabet.

TD-er committed:
Merge branch 'mega' into SetDatetime_1
